### PR TITLE
[new] chunk backend: Integrate Alibaba Cloud oss

### LINKED ISSUE #5499: Integrate Alibaba Cloud oss
Will there be any plans to integrate Alibaba Cloud's object storage oss?

### PR DESCRIPTION
**Pull request by liguozhong**
**What this PR does / why we need it**:
[new] chunk backend: Integrate Alibaba Cloud oss

**Which issue(s) this PR fixes**:
Fixes #5499 and #3504

**Special notes for your reviewer**:
yaml config
```yaml
schema_config:
  configs:
  - from: 2020-05-15
    object_store: oss
    schema: v11
    index:
      prefix: loki_index_
      period: 168h

storage_config:
  alibaba:
    bucket: loki-chunk-test
    endpoint: oss-cn-beijing.aliyuncs.com
    access_key_id: <ak>
    secret_access_key: <sk>
```

oss chunk by loki
image
debug oss getObject success.
image
query oss chunk data success.
image

**Checklist**
- [x] Documentation added
- [ ] Tests updated
- [x] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

**Comment by liguozhong** [comment]
> Hi! This issue has been automatically marked as stale because it has not had any activity in the past 30 days.
> 
> We use a stalebot among other tools to help manage the state of issues in this project. A stalebot can be very useful in closing issues in a number of cases; the most common is closing issues or PRs where the original reporter has not responded.
> 
> Stalebots are also emotionless and cruel and can close issues which are still very relevant.
> 
> **If this issue is important to you, please add a comment to keep it open. More importantly, please add a thumbs-up to the original issue entry.**
> 
> We regularly sort for closed issues which have a `stale` label sorted by thumbs up.
> 
> We may also:
> 
> * Mark issues as `revivable` if we think it's a valid issue but isn't something we are likely
>   to prioritize in the future (the issue will still remain closed).
> * Add a `keepalive` label to silence the stalebot if the issue is very common/popular/important.
> 
> We are doing our best to respond, organize, and prioritize all issues but it can be a challenging task, our sincere apologies if you find yourself at the mercy of the stalebot.

@stale ??

File: pkg/storage/chunk/client/alibaba/oss_object_client.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+// IsObjectNotFoundErr returns true if error means that object is not found. Relevant to GetObject and DeleteObject operations.
+func (s *OssObjectClient) IsObjectNotFoundErr(err error) bool {
+	return errors.Is(err, storage.ErrObjectNotExist)
```
Comment (owen-d) [inline pkg/storage/chunk/client/alibaba/oss_object_client.go:157]: This looks copied/pasted again and is comparing google errors to alibaba responses

Comment (liguozhong) [reply pkg/storage/chunk/client/alibaba/oss_object_client.go:157]: thanks.
The IsObjectNotFoundErr return value of error has been checked in detail.
At the same time, the test function has been added to confirm that it is running correctly.

ossErrorDeatil message snapshot.
image
test case pass. IsObjectNotFoundErr and GetObject
image

File: pkg/storage/chunk/client/alibaba/oss_object_client.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	"google.golang.org/api/option"
+
+	"github.com/grafana/loki/pkg/storage/chunk/client"
+)
+
+var ossRequestDuration = instrument.NewHistogramCollector(prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace: "loki",
+	Name:      "oss_request_duration_seconds",
+	Help:      "Time spent doing OSS requests.",
+	Buckets:   []float64{.025, .05, .1, .25, .5, 1, 2},
```
[the remark this answers is not in the document]
Comment (liguozhong) [reply pkg/storage/chunk/client/alibaba/oss_object_client.go:23]: thanks.done

File: pkg/storage/chunk/client/alibaba/oss_object_client.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	"github.com/aliyun/aliyun-oss-go-sdk/oss"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/instrument"
+	"google.golang.org/api/option"
+
+	"github.com/grafana/loki/pkg/storage/chunk/client"
+)
+
+var ossRequestDuration = instrument.NewHistogramCollector(prometheus.NewHistogramVec(prometheus.HistogramOpts{
```
[the remark this answers is not in the document]
Comment (liguozhong) [reply pkg/storage/chunk/client/alibaba/oss_object_client.go:19]: ok

File: pkg/storage/chunk/client/alibaba/oss_object_client.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	Namespace: "loki",
+	Name:      "oss_request_duration_seconds",
+	Help:      "Time spent doing OSS requests.",
+	Buckets:   []float64{.025, .05, .1, .25, .5, 1, 2},
+}, []string{"operation", "status_code"}))
+
+type ClientFactory func(ctx context.Context, opts ...option.ClientOption) (*storage.Client, error)
```
[the remark this answers is not in the document]
Comment (liguozhong) [reply pkg/storage/chunk/client/alibaba/oss_object_client.go:26]: thanks,I will delete ClientFactory

File: pkg/storage/chunk/client/alibaba/oss_object_client.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (liguozhong) [reply pkg/storage/chunk/client/alibaba/oss_object_client.go:19]: instrument.CollectedRequest can easily implement RED monitoring, I will continue to use `instrument` to be easier. The init() function has been registered

File: docs/sources/configuration/examples.md [hunk 5]
```diff
@@ -3,6 +3,29 @@ title: Examples
...
 
+## alibaba-cloud-storage-config.yaml
+
+```yaml
+# This partial configuration uses Alibaba for chunk storage
+
+schema_config:
+  configs:
+  - from: 2020-05-15
+    object_store: oss
```
[the remark this answers is not in the document]
Comment (liguozhong) [reply docs/sources/configuration/examples.md:14]: Great advice, the name `oss` of this product is really confusing. 
I have changed `oss` to `alibabacloud`, and `alibabacloud` has only one object storage, so there is no need to specify `oss` products.

File: pkg/storage/chunk/client/alibaba/oss_object_client.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	defaultBucket *oss.Bucket
+}
+
+// StorageConfig specifies config for storing data on AWS.
+type StorageConfig struct {
+	OssConfig `yaml:",inline"`
+}
+
+// OssConfig is config for the OSS Chunk Client.
+type OssConfig struct {
```
[the remark this answers is not in the document]
Comment (liguozhong) [reply pkg/storage/chunk/client/alibaba/oss_object_client.go:38]: Yes, the oss product of alibabacloud is a follower of aws s3, many fields are very similar

**Comment by liguozhong** [comment]
> With some tweaks, it looks like this should work. Have you tried it?

Yes, it can work normally when started directly in my local IDEA golang environment.

1:  It can be sent to the oss object storage of alibabacloud, 
2: and the logs can be found. At the same time, 
3: there are indicators of prometheus, 

which seems to be working. 

But I didn't deploy it to my production environment. If there is a related bug in the issue, I will fix it as soon as possible.

/metrics snapshot
image

oss console,loki data.
image

**Comment by owen-d** [comment]
Sounds good, thanks! I think we can reopen, fix the comments in my review, then merge :)

File: docs/sources/configuration/_index.md [hunk 7]
```diff
@@ -808,6 +808,29 @@ The `gcs_storage_config` configures GCS as a general storage for different data
...
+# Comma separated list of bucket names to evenly distribute chunks over.
+# CLI flag: -<prefix>.oss.buckets
+[bucket: <string> | default = ""]
+
+# oss Endpoint to connect to.
+# CLI flag: -<prefix>.oss.endpoint
+[endpoint: <string> | default = ""]
+
+# AWS Access Key ID
+# CLI flag: -<prefix>.s3.access-key-id
```
[the remark this answers is not in the document]
Comment (liguozhong) [reply docs/sources/configuration/_index.md:825]: ok,done.
image

**Comment by liguozhong** [comment]
> fix the comments

done. ready for review .



> I think we can reopen

This PR was automatically closed by robot, I don't have permission to reopen it

> then merge

that's great😃

**Comment by owen-d** [comment]
I can't reopen it either :( 
image


Would you mind creating a new PR and linking this one in the description?

**Comment by chengxiangwang** [comment]
chunks can't upload

**Comment by liguozhong** [comment]
> chunks can't upload

Can you file a new issue, follow the issue bug template, and provide more context such as loki version, loki configuration, loki error log, etc.
 Alibaba cloud oss storage works success in my environment. I can help you troubleshoot this problem

**Comment by yongzhang** [comment]
@liguozhong would you plz continue this PR to save lots of China users? I think it's OK for having potential issues, users like us can provide feedback and help improvements.

**Pull request closed without merging**